File: backend/Api/Master/Master.Entity/Models/MenuModels.cs
namespace Master.Entity.Models;

/// <summary>One module in the navigation rail, as the client receives it.</summary>
public class MenuView
{
    public int MenuId { get; set; }
    public string Code { get; set; } = null!;
    public string Name { get; set; } = null!;
    public string? Icon { get; set; }
    public string Module { get; set; } = null!;

    /// <summary>Set when the entry navigates directly instead of opening a panel.</summary>
    public string? RoutePath { get; set; }

    public bool IsSearchable { get; set; }
    public List<MenuGroupView> Groups { get; set; } = [];
}

/// <summary>A section within a module's panel. <see cref="Name"/> is null for a flat list.</summary>
public class MenuGroupView
{
    public int MenuGroupId { get; set; }
    public string Code { get; set; } = null!;
    public string? Name { get; set; }

    /// <summary>Lucide icon name for the section heading.</summary>
    public string? Icon { get; set; }

    public List<SubMenuView> SubMenus { get; set; } = [];
}

public class SubMenuView
{
    public int SubMenuId { get; set; }
    public string Code { get; set; } = null!;
    public string Name { get; set; } = null!;
    public string? RoutePath { get; set; }
    public string? Icon { get; set; }
    public string Module { get; set; } = null!;

    /// <summary>Whether this screen offers an inline create action at all.</summary>
    public bool CanCreate { get; set; }

    /// <summary>What one record is called, for the create button's label.</summary>
    public string? SingularName { get; set; }

    public bool HasAccess { get; set; }

    /// <summary>
    /// The actions this user holds on this screen — view, create, edit, delete,
    /// print, export, void, approve. A screen can hide what the role cannot do.
    /// </summary>
    public List<string> AllowedActions { get; set; } = [];
}

File: frontend/libs/app-shell/src/lib/menu.models.ts
/**
 * The navigation tree as `GET /api/menu` returns it.
 *
 * The server has already filtered this for the caller: a screen they hold no
 * permission on is gone, an empty group is gone, and a module left with nothing
 * to show is gone with it. Nothing here needs filtering again on the client — the
 * shape is what this user may actually reach.
 */
export interface MenuView {
  readonly menuId: number;
  readonly code: string;
  readonly name: string;
  readonly icon: string | null;
  readonly module: string;
  /** Set when the rail entry navigates directly instead of opening a panel. */
  readonly routePath: string | null;
  readonly isSearchable: boolean;
  readonly groups: readonly MenuGroupView[];
}

/** A section of a module's panel. `name` is null for a flat list. */
export interface MenuGroupView {
  readonly menuGroupId: number;
  readonly code: string;
  readonly name: string | null;
  /** Lucide icon name for the section heading. */
  readonly icon: string | null;
  readonly subMenus: readonly SubMenuView[];
}

export interface SubMenuView {
  readonly subMenuId: number;
  readonly code: string;
  readonly name: string;
  readonly routePath: string | null;
  readonly icon: string | null;
  readonly module: string;
  readonly canCreate: boolean;
  /** What one record is called — "Invoice" — for the create button's label. */
  readonly singularName: string | null;
  readonly hasAccess: boolean;
  /**
   * What this user may do here: view, create, edit, delete, print, export, void,
   * approve. A screen can hide the buttons the role cannot use.
   */
  readonly allowedActions: readonly string[];
}
